### PR TITLE
Fix rebuild index page compression

### PR DESCRIPTION
**Pull request by tongwu-msft**
## Description
Fix rebuild index issue, add data compression.

## Related issues
Addresses [issue #].

## Testing
Update integration tests

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [x] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImportOperation.cs [hunk 1]
```diff
@@ -81,64 +81,64 @@ public class SqlImportOperation : ISqlImportOperation, IImportOrchestratorTaskDa
...
+            list.Add((VLatest.TokenQuantityCompositeSearchParam, VLatest.TokenQuantityCompositeSearchParam.IX_TokenQuantityCompositeSearchParam_SearchParamId_Code1_QuantityCodeId2_SingleValue2, true));
+            list.Add((VLatest.TokenSearchParam, VLatest.TokenSearchParam.IX_TokenSeachParam_SearchParamId_Code_SystemId, true));
+            list.Add((VLatest.TokenStringCompositeSearchParam, VLatest.TokenStringCompositeSearchParam.IX_TokenStringCompositeSearchParam_SearchParamId_Code1_Text2, true));
+            list.Add((VLatest.TokenStringCompositeSearchParam, VLatest.TokenStringCompositeSearchParam.IX_TokenStringCompositeSearchParam_SearchParamId_Code1_Text2WithOverflow, true));
+            list.Add((VLatest.TokenText, VLatest.TokenText.IX_TokenText_SearchParamId_Text, true));
+            list.Add((VLatest.TokenTokenCompositeSearchParam, VLatest.TokenTokenCompositeSearchParam.IX_TokenTokenCompositeSearchParam_Code1_Code2, true));
+            list.Add((VLatest.UriSearchParam, VLatest.UriSearchParam.IX_UriSearchParam_SearchParamId_Uri, true));
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImportOperation.cs:142]: Ho do we ensure that this list matches schema definition in SQL scripts?

**Comment by SergeyGaluzo** [comment]
This fix should allow us to use disable/rebuild indexes on initial $import in PaaS. Right?

**Review by SergeyGaluzo — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql [hunk 2]
```diff
@@ -14,27 +14,37 @@
...
+       ,@isDisabled as bit = 0
+       ,@isExecuted as int = 0
+
+IF object_id(@tableName) IS NULL
+BEGIN
+    SET @errorTxt = @tableName +' does not exist or you don''t have permissions.'
+    RAISERROR(@errorTxt, 18, 127)
+END
+
+SELECT TOP 1 @isDisabled = is_disabled FROM sys.indexes WHERE object_id = object_id(@tableName) AND name = @indexName
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql:35]: Change to this SET @IsDisabled = (SELEC ...)

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql [hunk 3]
```diff
@@ -14,27 +14,37 @@
...
+END
+
+IF @isDisabled = 0
+BEGIN
+    SET @sql = N'ALTER INDEX ' + QUOTENAME(@indexName) + N' on ' + @tableName + ' Disable'
+    EXECUTE sp_executesql @sql
+    
+    SET @isExecuted = 1
+END
+RETURN @isExecuted
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql:49]: Why do you need to know this flag in C#?

**Review by SergeyGaluzo — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+--     @indexName
+--         * index name
+--     @pageCompression
+--         * index page compression
+
+GO
+CREATE PROCEDURE [dbo].[RebuildIndex_2]
+    @tableName nvarchar(128),
+    @indexName nvarchar(128),
+    @pageCompression bit
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:23]: Please apply the same comment from the DisableIndex

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+-- PARAMETERS
+--     @tableName
+--         * index table name
+--     @indexName
+--         * index name
+--     @pageCompression
+--         * index page compression
+
+GO
+CREATE PROCEDURE [dbo].[RebuildIndex_2]
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:20]: >[dbo].[RebuildIndex_2] [](http://example.com/codeflow?start=17&length=22)

Remove {}

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql [hunk 6]
```diff
@@ -14,27 +14,37 @@
 --         * index table name
 --     @indexName
 --         * index name
+
+GO
 CREATE PROCEDURE [dbo].[DisableIndex]
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql:19]: >[dbo].[DisableIndex] [](http://example.com/codeflow?start=17&length=20)

Remove []

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	BEGIN
+		SET @sql = N'ALTER INDEX ' + QUOTENAME(@indexName) + N' on ' + @tableName + ' Rebuild WITH (DATA_COMPRESSION = PAGE)'
+	END
+
+	EXECUTE sp_executesql @sql
+	SET @isExecuted = 1
+END
+        
+RETURN @isExecuted
+GO
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:60]: Please remove.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+CREATE PROCEDURE dbo.RebuildIndex_2
+    @tableName nvarchar(128),
+    @indexName nvarchar(128),
+    @pageCompression bit
+WITH EXECUTE AS 'dbo'
+AS
+DECLARE @errorTxt as varchar(1000)
+       ,@sql as nvarchar (1000)
+       ,@isDisabled as bit = 0
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:28]: > = 0 [](http://example.com/codeflow?start=26&length=4)

no need to set it.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImportOperation.cs (hunk 1 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImportOperation.cs:142]: Thanks, add integration test to check the compression not change after disable/rebuild.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql (hunk 2 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql]: Thanks, updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql (hunk 3 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql]: Currently we do not use it in product code but in test we use this to make sure the if rebuild on the index not disabled should skip. And remove the flag for disable index.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 4 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:23]: thanks, updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql (hunk 6 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/DisableIndex.sql]: thanks, updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 5 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql]: thanks. updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 7 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:60]: Currently we do not use it in product code but in test we use this to make sure the if rebuild on the index not disabled should skip. And remove the flag for disable index.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 8 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql]: Thanks, updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+--     @indexName
+--         * index name
+--     @pageCompression
+--         * index page compression
+
+GO
+CREATE PROCEDURE dbo.RebuildIndex_2
+    @tableName nvarchar(128),
+    @indexName nvarchar(128),
+    @pageCompression bit
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:23]: >@pageCompression bit [](http://example.com/codeflow?start=4&length=20)

Please add default = 0 and use previous RebuildIndex stored procedure (no need to add new one).

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 7 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql:60]: Hmm... 

So you don't trust stored procedure to skip index rebuild correctly but trust it to set flag on the same condition? 

This sounds incorrect.

Please proceed with removing this flag. You may need to remove related to this flag test as well.

If you need help in writing test that truly checks whether index was rebuild please let me know.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 9 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql]: Thanks, updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql (hunk 7 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/RebuildIndex_2.sql]: Remove this flag, and later we can apply new rebuild index logic.

**Review by SergeyGaluzo — APPROVED** [review]
:shipit:

**Review by feordin — APPROVED** [review]

**Review by LTA-Thinking — APPROVED** [review]

**Pull request merged**